src/main/kotlin/org/rust/ide/intentions/DemorgansLawIntention.kt

```diff
@@ -0,0 +1,107 @@
+package org.rust.ide.intentions
+
+import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.tree.IElementType
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.util.elementType
+import org.rust.lang.core.psi.util.parentOfType
+
+class DemorgansLawIntention : PsiElementBaseIntentionAction() {
+    override fun getText() = "DeMorgan Law"
+    override fun getFamilyName() = text
+    override fun startInWriteAction() = true
+
+    private data class Context(
+        val binaryExpr: RustBinaryExprElement,
+        val binaryOpType: IElementType
+    )
+
+    private fun findContext(element: PsiElement): Context? {
+        if (!element.isWritable) return null
+
+        val binExpr = element.parentOfType<RustBinaryExprElement>() ?: return null
+        val opType = binExpr.operatorType
+        if (opType == RustTokenElementTypes.ANDAND || opType == RustTokenElementTypes.OROR) {
+            return Context(binExpr, opType)
+        }
+        return null
+    }
+
+    override fun invoke(project: Project, editor: Editor?, element: PsiElement) {
+        val ctx = findContext(element) ?: return
+        var binExpr = ctx.binaryExpr
+        val opType = ctx.binaryOpType
```

Comment:
Not exactly; in Kotlin this is just a sugar:
```kotlin
val (a, b, c) = Foo()
```
=>
```kotlin
val __foo = Foo()
val a = __foo.component1()
val b = __foo.component2()
val c = __foo.component3()
```

Not so powerful as in ML/Rust/Haskell/Scala, but still pretty neat